awx/main/notifications/grafana_backend.py

```diff
@@ -0,0 +1,63 @@
+# Copyright (c) 2016 Ansible, Inc.
+# All Rights Reserved.
+
+import logging
+import requests
+import dateutil.parser as dp
+
+from django.utils.encoding import smart_text
+from django.utils.translation import ugettext_lazy as _
+from base import AWXBaseEmailBackend
+
+logger = logging.getLogger('awx.main.notifications.grafana_backend')
+
+
+class GrafanaBackend(AWXBaseEmailBackend):
+
+    init_parameters = {"grafana_url": {"label": "Grafana URL", "type": "string"},
+                       "grafana_key": {"label": "Grafana API Key", "type": "password"}}
+    recipient_parameter = "grafana_url"
+    sender_parameter = None
+
+    def __init__(self, grafana_key,dashboardId=None, panelId=None, annotation_tags=None, grafana_no_verify_ssl=False, isRegion=True,
+                 fail_silently=False, **kwargs):
+        super(GrafanaBackend, self).__init__(fail_silently=fail_silently)
+        self.grafana_key = grafana_key
+        self.dashboardId = dashboardId
+        self.panelId = panelId
+        self.annotation_tags = annotation_tags if annotation_tags is not None else []
+        self.grafana_no_verify_ssl = grafana_no_verify_ssl
+        self.isRegion = isRegion
+
+    def format_body(self, body):
+        return body
+
+    def send_messages(self, messages):
+        sent_messages = 0
+        for m in messages:
+            grafana_data = {}
+            grafana_headers = {}
+            try:
+                grafana_data['time'] = int(dp.parse(m.body['started']).strftime('%s%f')[:-3])
```

Comment:
`.strftime("%s")` is not really portable, as support for the `%s` format code is incidental and not an actual feature of Python's implementation.

Does something like this work w/ Graphana's API?

```python
(dp.parse(x) - datetime(1970, 1, 1)).total_seconds()
1546429153.3778
```

Reply:
Many thanks for your review and findings. I thought about something like this, what do you think? 

```
epoch=datetime.datetime.utcfromtimestamp(0)

grafana_data['time'] = int((dp.parse(m.body['started']).replace(tzinfo=None)-epoch).total_seconds()*1000)

grafana_data['timeEnd'] = int((dp.parse(m.body['finished']).replace(tzinfo=None)-epoch).total_seconds()*1000)
```


Either i remove the timezone information from the parsed body parameters with tzinfo=None or i add timezone information to epoch (but this timezone information must then be equal to the one delivered in the body parameters which would make the code more complicated from my point of view).

Reply:
```python
>>> x = datetime.utcnow().isoformat()
>>> (dp.parse(x) - datetime(1970, 1, 1)).total_seconds()
1546610339.866411
>>> int((dp.parse(x).replace(tzinfo=None)-datetime.utcfromtimestamp(0)).total_seconds()*1000)
1546610339866
```

Yep, this approach makes sense to me - my main concern was that `%s` is a little esoteric and _could_ cause compatibility problems in the future.

`body["started"]` and `body["finished"]` are generated here: https://github.com/ansible/awx/blob/devel/awx/main/models/unified_jobs.py#L1212-L1213

...which in turn are `datetime.datetime` objects that come from these two ORM fields:

https://github.com/ansible/awx/blob/devel/awx/main/models/unified_jobs.py#L625-L636

...so they're going to be datetimes w/ tzinfo=utc.  So I'd say making the dates naive with `tzinfo=None` should be fine here.

Have you tested a notification with this code and confirmed that it shows the local/appropriate time you'd expect in Grafana?

Reply:
Ok great, thanks for the fast response 👍 I agree that %s is not the best option. I tested a notification with this code already and it works just like before. Depending on the configuration in Grafana Settings regarding timezone ("Local browser time" or "UTC") the annotation shows the correct time. Should i commit the code or is there anything else that needs some review? Many thanks again so far.

Reply:
@crab86 at this point, I'm happy with the Python code - I'm going to get one of our JS contributors to review the front-end side, and we'll probably do some testing on our end.  It may be a bit while we decide if and when to merge this milestone-wise, but _I_ think this is a pretty reasonable addition, and I think this PR looks good if you'll update this epoch code as discussed.

Reply:
Definately willing to do so, but may take some days - lacking experience on mock tests.

Reply:
@ryanpetrello I just added unit tests, since those are my first mock tests it would be great if you could review them to see if they fit your needs or if they need to be adapted.